dvc/repo/run.py

```diff
@@ -33,30 +38,34 @@ def _get_file_path(**kwargs):
 @scm_context
 def run(self, fname=None, no_exec=False, **kwargs):
     from dvc.stage import PipelineStage, Stage, create_stage
-    from dvc.dvcfile import DVC_FILE, Dvcfile
+    from dvc.dvcfile import Dvcfile, PIPELINE_FILE
 
-    stage_cls, path = PipelineStage, fname or DVC_FILE
-    if not kwargs.get("name"):
+    stage_cls, path = PipelineStage, PIPELINE_FILE
+    stage_name = kwargs.get("name")
+    if not stage_name:
         kwargs.pop("name", None)
-        stage_cls, path = Stage, fname or _get_file_path(**kwargs)
+        stage_cls, path = Stage, fname or _get_file_path(kwargs)
+    else:
+        if not is_valid_name(stage_name):
+            raise InvalidStageName
 
     stage = create_stage(stage_cls, repo=self, path=path, **kwargs)
     if stage is None:
         return None
 
-    dvcfile = Dvcfile(self, path)
-    if dvcfile.exists() and not dvcfile.is_multi_stage():
-        if stage_cls == PipelineStage:
-            raise DvcException(
-                "'{}' is a single-stage dvcfile. Please use "
-                "`-f <different-filename>` and try again.`.".format(
-                    relpath(dvcfile.path)
-                )
-            )
-        dvcfile.remove_with_prompt(force=kwargs.get("overwrite", True))
-
-    self.check_modified_graph([stage], self.pipeline_stages)
+    dvcfile = Dvcfile(self, stage.path)
+    if dvcfile.exists():
+        if stage_name and stage_name in dvcfile.stages:
+            raise DuplicateStageName(stage_name, dvcfile)
+        if stage_cls != PipelineStage:
+            dvcfile.remove_with_prompt(force=kwargs.get("overwrite", True))
+
+    try:
+        self.check_modified_graph([stage])
+    except OutputDuplicationError as exc:
+        raise OutputDuplicationError(exc.output, set(exc.stages) - {stage})
```

Comment:
```suggestion
        raise OutputDuplicationError(exc.output, set(exc.stages) - {stage}) from exc
```
maybe?

Reply:
Also, not sure I understand why we need to remove stage from exc.stages. We didn't do that before , did we?

Reply:
The stage that we are creating is also shown in the error message:
```sh
ERROR: file/directory 'a2.txt' is specified as an output in more than one stage:
        a2.txt.dvc
	pipelines.yaml:copy-a2-txt
```

Reply:
@skshetry That looks alright, or am I missing something?

Reply:
The thing is `a2.txt.dvc` does not exist at this point. So, I find it to be a bit misleading.